{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cooltools import snipping\n",
    "import cooler\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import bioframe\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import glob\n",
    "import os\n",
    "import multiprocess\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 16})\n",
    "chromsizes=bioframe.fetch_chromsizes('mm10')\n",
    "chromosomes = list(chromsizes.index)\n",
    "binsize=5000\n",
    "flank = 300000\n",
    "supports = [(chrom, 0, chromsizes[chrom]) for chrom in chromosomes]\n",
    "binsDf=pd.read_csv('../data/genomeInfo/mm10bin5kb.bed',sep='\\t',header=None,names=['chrom','start','end']) #mm10bin5kb.bed file make using cooler makebins\n",
    "signalFiles=glob.glob('../data/binnedBedgraphs/*')\n",
    "signalDf=binsDf.copy()\n",
    "for signalFile in signalFiles:\n",
    "    signalDf=signalDf.merge(pd.read_csv(signalFile,sep=\"\\t\", header=None,names=['chrom','start','end',os.path.basename(signalFile).split('.')[0]]),how='left',on=['chrom','start','end'])\n",
    "    signalDf.drop_duplicates(subset=['chrom','start','end'],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getPeakWindowsDf(peakFile,binsize=5000,flank=300000,supports=supports,binsDf=binsDf):\n",
    "    peakDf=pd.read_csv(peakFile,header=None,sep=\"\\t\",names=['chrom','start','end','name','score','strand','signalValue','pvalue','qvalue','peak'])\n",
    "    windows = snipping.make_bin_aligned_windows(\n",
    "        binsize,\n",
    "        peakDf['chrom'],\n",
    "        (peakDf['start'] + peakDf['end'])//2,\n",
    "        flank_bp=flank)\n",
    "    windows=windows.merge(pd.DataFrame(supports,columns=['chrom','zero','chromsize']),how='left',on='chrom')\n",
    "    windows=windows[(windows.end<windows.chromsize) & (windows.start>0)].drop_duplicates()\n",
    "    windows=windows.loc[(windows.chrom!='chrX')&(windows.chrom!='chrY')&(windows.chrom!='chrM'),:].reset_index(drop=True)\n",
    "    startInds=binsDf.reset_index().merge(windows[['chrom','start']],how='right',on=['chrom','start'])['index']\n",
    "    endInds=binsDf.reset_index().merge(windows[['chrom','end']],how='right',on=['chrom','end'])['index']\n",
    "    windows['binID']=(endInds+startInds)//2\n",
    "    return windows\n",
    "def getPeakSnipWindows(peakFile,binsize=5000,flank=300000,supports=supports,binsDf=binsDf):\n",
    "    if type(peakFile) is str:\n",
    "        peakDf=pd.read_csv(peakFile,header=None,sep=\"\\t\",names=['chrom','start','end','name','score','strand','signalValue','pvalue','qvalue','peak'])\n",
    "    elif type(peakFile) is pd.core.frame.DataFrame:\n",
    "        peakDf=peakFile\n",
    "    windows = snipping.make_bin_aligned_windows(\n",
    "        binsize,\n",
    "        peakDf['chrom'],\n",
    "        (peakDf['start'] + peakDf['end'])//2,\n",
    "        flank_bp=flank)\n",
    "    windows=snipping.assign_regions(windows,supports=supports).drop_duplicates().reset_index(drop=True)\n",
    "    windows=windows.loc[(windows.chrom!='chrX')&(windows.chrom!='chrY')&(windows.chrom!='chrM'),:].reset_index(drop=True)\n",
    "    windows['binID']=binsDf.reset_index().merge(windows[['chrom','start']],how='right',on=['chrom','start'])['index']+flank//binsize\n",
    "    return windows\n",
    "def splitQuantile(toSplit,splitOn,quantiles=4):\n",
    "    quantileLims=np.linspace(0,len(toSplit),quantiles+1).astype(int)\n",
    "    sortedArray=splitOn[toSplit].sort_values().index.to_numpy()\n",
    "    quantiles=[sortedArray[quantileLims[ind]:quantileLims[ind+1]] for ind in range(quantiles)]\n",
    "    return quantiles\n",
    "def diagAverageMat(inputMat):\n",
    "    imSize=inputMat.shape[0]\n",
    "    diagAverages=np.zeros(imSize*2-1)\n",
    "    diagAverageMat=np.zeros((imSize,imSize))\n",
    "    for diagInd in range(2*imSize-1):\n",
    "        meanDiagVec=np.nanmean(np.diag(inputMat,diagInd-imSize+1))*np.ones(np.min([diagInd, 2*imSize-2-diagInd])+1)\n",
    "        diagAverageMat=diagAverageMat+np.diag(meanDiagVec, diagInd-imSize+1)\n",
    "    return diagAverageMat\n",
    "\n",
    "def getPairedSnipWindows(snipWindows1,snipWindows2=None,maxDist=1000000,minDist=400000,ordered=True):\n",
    "    if snipWindows2 is None:\n",
    "        snipWindows2=snipWindows1\n",
    "    df=snipWindows1.merge(snipWindows2,on='region',suffixes=[1,2])\n",
    "    if ordered:\n",
    "        return df[(np.abs(df.start1-df.start2)<=maxDist)&(np.abs(df.start1-df.start2)>=minDist)&(df.start1<df.start2)]\n",
    "    else:\n",
    "        return df[(np.abs(df.start1-df.start2)<=maxDist)&(np.abs(df.start1-df.start2)>=minDist)]\n",
    "\n",
    "def plotSnipPileup(snipWindows,coolerHandle,ax,vmin=0.95,vmax=1.05,binsize=5000,flank=300000,titleString=\"\",nthreads=8):\n",
    "    snipper = snipping.CoolerSnipper(coolerHandle)\n",
    "    with multiprocess.Pool(nthreads) as pool:\n",
    "        stack = snipping.pileup(\n",
    "            snipWindows,\n",
    "            snipper.select,\n",
    "            snipper.snip,\n",
    "            map=pool.map)\n",
    "    avgMat=np.nanmean(stack, axis=2)\n",
    "    oeMat=avgMat/diagAverageMat(avgMat)\n",
    "    extent=[-flank/1000,flank/1000,flank/1000,-flank/1000]\n",
    "    plt.colorbar(mappable=ax.imshow(oeMat,extent=extent, cmap='coolwarm',vmin=vmin, vmax=vmax),\n",
    "                         ax=ax, shrink=0.6, label='fold enrichment')\n",
    "    ax.set_aspect('equal')\n",
    "    ax.set_xlabel('distance to peak (kb)')\n",
    "    ax.set_ylabel('distance to peak (kb)')\n",
    "    ax.set_title(titleString)\n",
    "    \n",
    "def plotSnipPileups(snipWindowsDict,coolerDict,vmin=0.95,vmax=1.05,saveName=None):\n",
    "    fig,axs=plt.subplots(len(snipWindowsDict),len(coolerDict),figsize=(5*len(coolerDict),5*len(snipWindowsDict)))\n",
    "    for snipWindowsInd,snipWindowsKey in enumerate(snipWindowsDict):\n",
    "        for coolerInd,coolerDictKey in enumerate(coolerDict):\n",
    "            plotSnipPileup(snipWindowsDict[snipWindowsKey],coolerDict[coolerDictKey],axs[snipWindowsInd,coolerInd],\n",
    "                           titleString=coolerDictKey+\"\\n\"+snipWindowsKey,vmin=vmin, vmax=vmax)\n",
    "    fig.tight_layout()\n",
    "    if saveName!=None:\n",
    "        plt.savefig(saveName)\n",
    "    \n",
    "def plotPileup(windowsDf,coolerHandle,ax,vmin=0.95,vmax=1.05,binsize=5000,flank=300000,titleString=\"\"):\n",
    "    windowList=[(windowsDf.loc[ind,'chrom'],windowsDf.loc[ind,'start'],windowsDf.loc[ind,'end']) for ind in windowsDf.index]\n",
    "    mats=np.asarray([coolerHandle.matrix().fetch(window) for window in windowList])\n",
    "    avgMat=np.nanmean(mats, axis=0)\n",
    "    oeMat=avgMat/diagAverageMat(avgMat)\n",
    "    extent=[-flank/1000,flank/1000,flank/1000,-flank/1000]\n",
    "    plt.colorbar(mappable=ax.imshow(oeMat,extent=extent, cmap='coolwarm',vmin=0.95, vmax=1.05),\n",
    "                         ax=ax, shrink=0.6, label='fold enrichment')\n",
    "    ax.set_aspect('equal')\n",
    "    ax.set_xlabel('distance to peak (kb)')\n",
    "    ax.set_ylabel('distance to peak (kb)')\n",
    "    ax.set_title(titleString)\n",
    "\n",
    "    \n",
    "def plotPileups(windowListDict,coolerDict,vmin=0.95,vmax=1.05,saveName=None):\n",
    "\n",
    "    fig,axs=plt.subplots(len(windowListDict),len(coolerDict),figsize=(5*len(coolerDict),5*len(windowListDict)))\n",
    "    for windowListInd,windowListKey in enumerate(windowListDict):\n",
    "        for coolerInd,coolerDictKey in enumerate(coolerDict):\n",
    "            plotPileup(windowListDict[windowListKey],coolerDict[coolerDictKey],axs[windowListInd,coolerInd],\n",
    "                       titleString=coolerDictKey+\"\\n\"+windowListKey,vmin=vmin, vmax=vmax)\n",
    "    fig.tight_layout()\n",
    "    if saveName!=None:\n",
    "        plt.savefig(saveName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peakFiles=glob.glob('../data/peakBeds/*')\n",
    "peakWindowsDfDict=dict(zip([os.path.basename(file).split('.')[0] for file in peakFiles],[getPeakWindowsDf(file) for file in peakFiles]))\n",
    "peakSnipWindowsDfDict=dict(zip([os.path.basename(file).split('.')[0] for file in peakFiles],[getPeakSnipWindows(file) for file in peakFiles]))\n",
    "\n",
    "saveDir='../outputs/'\n",
    "\n",
    "coolerListDict={\n",
    "    'ES':cooler.Cooler('../data/coolers/ES.mcool::resolutions/5000'), #From Bonev et al 2017\n",
    "    'zygonema':cooler.Cooler('../data/coolers/zygonema.mcool::resolutions/5000'), #from Patel et al 2019\n",
    "    'pachynema':cooler.Cooler('../data/coolers/pachynema.mcool::resolutions/5000'), #from Patel et al 2019\n",
    "}\n",
    "\n",
    "coolerListDictWPAC={\n",
    "    'ES':cooler.Cooler('../data/coolers/ES.mcool::resolutions/5000'), #From Bonev et al 2017\n",
    "    'zygonema':cooler.Cooler('../data/coolers/zygonema.mcool::resolutions/5000'), #from Patel et al 2019\n",
    "    'pachynema':cooler.Cooler('../data/coolers/pachynema.mcool::resolutions/5000'), #from Patel et al 2019\n",
    "    'PAC':cooler.Cooler('../data/coolers/PAC.mcool::resolutions/5000'), #from Wang et al 2019\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prdm9SplitonDMC=splitQuantile(peakSnipWindowsDfDict['Baker2015_prdm9_B6xCAST_peaks'].binID,signalDf.smagulova2016_B6xCAST_DSB_fraglen1000)\n",
    "prdm9WindowDict={'PRDM9 peaks (all)':peakSnipWindowsDfDict['Baker2015_prdm9_B6xCAST_peaks'],\n",
    "                   'PRDM9 peaks (top-DSB)':peakSnipWindowsDfDict['Baker2015_prdm9_B6xCAST_peaks'][peakSnipWindowsDfDict['Baker2015_prdm9_B6xCAST_peaks'].binID.isin(prdm9SplitonDMC[-1])],\n",
    "                   'PRDM9 peaks (bottom-DSB)':peakSnipWindowsDfDict['Baker2015_prdm9_B6xCAST_peaks'][peakSnipWindowsDfDict['Baker2015_prdm9_B6xCAST_peaks'].binID.isin(prdm9SplitonDMC[0])]}\n",
    "\n",
    "dsbSplitonCO=splitQuantile(peakSnipWindowsDfDict['smagulova2016_B6xCAST_DSB_fraglen1000_peaks'].binID,signalDf.yin2019_crossoversCast1C_binned_weightedScore_medianNormalized)\n",
    "dsbWindowDict={'DSB peaks (all)':peakSnipWindowsDfDict['smagulova2016_B6xCAST_DSB_fraglen1000_peaks'],\n",
    "                   'DSB peaks (top-CO)':peakSnipWindowsDfDict['smagulova2016_B6xCAST_DSB_fraglen1000_peaks'][peakSnipWindowsDfDict['smagulova2016_B6xCAST_DSB_fraglen1000_peaks'].binID.isin(dsbSplitonCO[-1])],\n",
    "                   'DSB peaks (bottom-CO)':peakSnipWindowsDfDict['smagulova2016_B6xCAST_DSB_fraglen1000_peaks'][peakSnipWindowsDfDict['smagulova2016_B6xCAST_DSB_fraglen1000_peaks'].binID.isin(dsbSplitonCO[0])]}\n",
    "\n",
    "cohesinWindowDict={'Meiotic RAD21L sites':peakSnipWindowsDfDict['vara2019_RSrad21l_peaks'],\n",
    "                          'Meiotic REC8 sites':peakSnipWindowsDfDict['vara2019_RSrec8_peaks'],\n",
    "                          'ES RAD21 sites':peakSnipWindowsDfDict['Nitzsche2011_ESC_RAD21_peaks']}\n",
    "\n",
    "cohesinPairWindowDict={'Meiotic RAD21L site-pairs':getPairedSnipWindows(peakSnipWindowsDfDict['vara2019_RSrad21l_peaks']),\n",
    "                          'Meiotic REC8 site-pairs':getPairedSnipWindows(peakSnipWindowsDfDict['vara2019_RSrec8_peaks']),\n",
    "                          'ES RAD21 site-pairs':getPairedSnipWindows(peakSnipWindowsDfDict['Nitzsche2011_ESC_RAD21_peaks'])}\n",
    "\n",
    "rad21lPosNegWindowDict={'RAD21L +strand TSS':peakSnipWindowsDfDict['vara2019_PDrad21l_peaks_posStrandUnique'],\n",
    "                          'RAD21L -strand TSS':peakSnipWindowsDfDict['vara2019_PDrad21l_peaks_negStrandUnique']}\n",
    "\n",
    "rad21lConvDivPairWindowDict={'RAD21L conv-pairs':getPairedSnipWindows(peakSnipWindowsDfDict['vara2019_PDrad21l_peaks_posStrandUnique'],peakSnipWindowsDfDict['vara2019_PDrad21l_peaks_negStrandUnique']),\n",
    "                          'RAD21L div-pairs':getPairedSnipWindows(peakSnipWindowsDfDict['vara2019_PDrad21l_peaks_negStrandUnique'],peakSnipWindowsDfDict['vara2019_PDrad21l_peaks_posStrandUnique'])}\n",
    "\n",
    "ctcfPosNegWindowDict={'CTCF +strand motif':peakSnipWindowsDfDict['Nitzsche2011_ESC_CTCF_peaks_posStrandUnique'],\n",
    "                          'CTCF -strand motif':peakSnipWindowsDfDict['Nitzsche2011_ESC_CTCF_peaks_negStrandUnique']}\n",
    "\n",
    "ctcfConvDivPairWindowDict={'CTCF conv-pairs':getPairedSnipWindows(peakSnipWindowsDfDict['Nitzsche2011_ESC_CTCF_peaks_posStrandUnique'],peakSnipWindowsDfDict['Nitzsche2011_ESC_CTCF_peaks_negStrandUnique']),\n",
    "                          'CTCF div-pairs':getPairedSnipWindows(peakSnipWindowsDfDict['Nitzsche2011_ESC_CTCF_peaks_negStrandUnique'],peakSnipWindowsDfDict['Nitzsche2011_ESC_CTCF_peaks_posStrandUnique'])}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "windowDictMetaList=[prdm9WindowDict,dsbWindowDict]\n",
    "windowKeys=['prdm9','dsb']\n",
    "windowDictMetaDict=dict(zip(windowKeys,windowDictMetaList))\n",
    "for key in windowDictMetaDict:\n",
    "    plotSnipPileups(snipWindowsDict=windowDictMetaDict[key],coolerDict=coolerListDictWPAC,saveName=f'{saveDir}/{key}_metahic.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "windowDictMetaList=[cohesinWindowDict,cohesinPairWindowDict,rad21lPosNegWindowDict,rad21lConvDivPairWindowDict,ctcfPosNegWindowDict,ctcfConvDivPairWindowDict]\n",
    "windowKeys=['cohesin','cohesinPairs','rad21lPosNeg','rad21lConvDiv','ctcfPosNeg','ctcfConvDiv']\n",
    "windowDictMetaDict=dict(zip(windowKeys,windowDictMetaList))\n",
    "for key in windowDictMetaDict:\n",
    "    plotSnipPileups(snipWindowsDict=windowDictMetaDict[key],coolerDict=coolerListDictWPAC,vmin=0.75,vmax=1.25,saveName=f'{saveDir}/{key}_metahic.pdf')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
