{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from scipy.signal import find_peaks\n",
    "from scipy import stats\n",
    "from patsy import dmatrices, dmatrix\n",
    "import statsmodels.api as sm\n",
    "import glob\n",
    "import os\n",
    "import bioframe\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import cooler\n",
    "from cooltools import snipping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 24})\n",
    "chromsizes=bioframe.fetch_chromsizes('mm10')\n",
    "chromosomes = list(chromsizes.index)\n",
    "binsize=5000\n",
    "flank = 300000\n",
    "supports = [(chrom, 0, chromsizes[chrom]) for chrom in chromosomes]\n",
    "binsDf=pd.read_csv('../data/genomeInfo/mm10bin5kb.bed',sep='\\t',header=None,names=['chrom','start','end']) #bin file can be made with cooler makebins, binsize = 5000\n",
    "\n",
    "signalFiles=glob.glob('../data/binnedBedgraphs/*')\n",
    "signalDf=binsDf.copy()\n",
    "for signalFile in signalFiles:\n",
    "    signalDf=signalDf.merge(pd.read_csv(signalFile,sep=\"\\t\", header=None,names=['chrom','start','end',os.path.basename(signalFile).split('.')[0]]),how='left',on=['chrom','start','end'])\n",
    "    signalDf.drop_duplicates(subset=['chrom','start','end'],inplace=True)\n",
    "varList=signalDf.columns[3:]\n",
    "rollSignalDf=signalDf.groupby(['chrom'])[varList].rolling(window=101, center=True, min_periods=1).mean().reset_index(level=0).drop(columns=['chrom'])\n",
    "signalDf=pd.merge(signalDf,rollSignalDf, left_index=True,right_index=True, suffixes=('', '_roll101mean'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getPeakInds(peakFile,binsize=5000,flank=300000,supports=supports):\n",
    "    peakDf=pd.read_csv(peakFile,header=None,sep=\"\\t\",names=['chrom','start','end','name','score','strand','signalValue','pvalue','qvalue','peak'])\n",
    "    windows = snipping.make_bin_aligned_windows(\n",
    "        binsize,\n",
    "        peakDf['chrom'],\n",
    "        (peakDf['start'] + peakDf['end'])//2,\n",
    "        flank_bp=flank)\n",
    "    windows=windows.merge(pd.DataFrame(supports,columns=['chrom','zero','chromsize']),how='left',on='chrom')\n",
    "    windows=windows[(windows.end<windows.chromsize) & (windows.start>0)].drop_duplicates()\n",
    "    windows=windows.loc[(windows.chrom!='chrX')&(windows.chrom!='chrY')&(windows.chrom!='chrM'),:]\n",
    "    startInds=binsDf.reset_index().merge(windows[['chrom','start']],how='inner',on=['chrom','start'])['index']\n",
    "    endInds=binsDf.reset_index().merge(windows[['chrom','end']],how='inner',on=['chrom','end'])['index']\n",
    "    indDf=pd.DataFrame({'startInd':startInds,'endInd':endInds})\n",
    "    return ((indDf.startInd+indDf.endInd)//2).values\n",
    "\n",
    "def plotCorrMatrix(df, corrMethod='pearson', saveName=None,title=None):\n",
    "    correlationsDf=df.fillna(0).corr(method=corrMethod)\n",
    "    # Generate a mask for the upper triangle\n",
    "    mask = np.triu(np.ones_like(correlationsDf, dtype=np.bool))\n",
    "\n",
    "    # Set up the matplotlib figure\n",
    "    f, ax = plt.subplots(figsize=(5+correlationsDf.shape[1], 1+correlationsDf.shape[1]))\n",
    "    # Generate a custom diverging colormap\n",
    "#     cmap = sns.diverging_palette(220, 0, as_cmap=True)\n",
    "\n",
    "    # Draw the heatmap with the mask and correct aspect ratio\n",
    "    sns.heatmap(correlationsDf, mask=mask, cmap=\"vlag\", vmax=.3, center=0,\n",
    "                square=True, linewidths=.5, cbar_kws={\"shrink\": .5,\"label\": corrMethod+' correlation'})\n",
    "    if title is not None:\n",
    "        plt.title(title)\n",
    "    if saveName is not None:\n",
    "        plt.savefig(saveName)\n",
    "    return correlationsDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def OLSBackwardSelectStep(X, y, selected, sigThreshold):\n",
    "    pVals=sm.OLS(y,X[:, selected]).fit().pvalues\n",
    "    if pVals.max()>sigThreshold:\n",
    "        selected[np.argwhere(selected)[np.argmax(pVals)]]=False\n",
    "        converged=False\n",
    "    else:\n",
    "        converged=True\n",
    "    return selected, converged\n",
    "    \n",
    "def OLSForwardSelectStep(X, y, selected, sigThreshold):\n",
    "    minMaxPVal=1\n",
    "    for ind in np.argwhere(~selected):\n",
    "        newSelectedCandidate=np.copy(selected)\n",
    "        newSelectedCandidate[ind]=True\n",
    "        maxPVal=sm.OLS(y,X[:,newSelectedCandidate]).fit().pvalues.max()\n",
    "        if maxPVal<minMaxPVal:\n",
    "            newSelected=newSelectedCandidate\n",
    "            minMaxPVal=maxPVal\n",
    "    if minMaxPVal>sigThreshold:\n",
    "        converged=True\n",
    "    else:\n",
    "        selected=newSelected\n",
    "        converged=False\n",
    "    return selected, converged\n",
    "\n",
    "def OLSBackwardSelect(X, y, sigThreshold):\n",
    "    converged=False\n",
    "    selected=np.ones(X.shape[1]).astype('bool')\n",
    "    while (np.sum(selected)>1) & (not converged):\n",
    "        selected, converged = OLSBackwardSelectStep(X, y, selected=selected, sigThreshold=sigThreshold)\n",
    "    return selected\n",
    "\n",
    "def OLSForwardSelect(X, y, sigThreshold):\n",
    "    converged=False\n",
    "    selected=np.zeros(X.shape[1]).astype('bool')\n",
    "    selected[0]=True\n",
    "    while (np.sum(selected)<X.shape[1]) & (not converged):\n",
    "        selected, converged = OLSForwardSelectStep(X, y, selected=selected, sigThreshold=sigThreshold)\n",
    "    return selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peakFiles=glob.glob('../data/peakBeds/*')\n",
    "peakIndsDict=dict(zip([os.path.basename(file).split('.')[0] for file in peakFiles],[getPeakInds(file) for file in peakFiles]))\n",
    "saveDir='../outputs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictorList=['ES_cisTotalRatio',\n",
    "            'zygo_cisTotalRatio',\n",
    "            'pachy_cisTotalRatio',\n",
    "            'ES_E1chunks',\n",
    "            'zygo_E1chunks',\n",
    "            'pachy_E1chunks',\n",
    "            'ES_fireScore',\n",
    "            'zygo_fireScore',\n",
    "            'pachy_fireScore',\n",
    "            'ES_logInsScore',\n",
    "            'zygo_logInsScore',\n",
    "            'pachy_logInsScore',\n",
    "            'chromHMMstate_1_binned',\n",
    "            'chromHMMstate_2_binned',\n",
    "            'chromHMMstate_3_binned',\n",
    "            'chromHMMstate_4_binned',\n",
    "            'chromHMMstate_5_binned',\n",
    "            'chromHMMstate_6_binned',\n",
    "            'chromHMMstate_7_binned',\n",
    "            'vara2019_PDrad21l',\n",
    "            'vara2019_PDctcf',\n",
    "            'margolin2014RNAPII16dpp',\n",
    "            'Nitzsche2011_ESC_RAD21',\n",
    "            'Nitzsche2011_ESC_CTCF',\n",
    "            'ES_RNAP2_binned_maxScore',\n",
    "]\n",
    "simpleVarNames=['ES cis-total ratio',\n",
    "                'Zygonema cis-total ratio',\n",
    "                'Pachynema cis-total ratio',\n",
    "                'ES compartment vector',\n",
    "                'Zygonema compartment vector',\n",
    "                'Pachynema compartment vector',\n",
    "                'ES FIRE score',\n",
    "                'Zygonema FIRE score',\n",
    "                'Pachynema FIRE score',\n",
    "                'ES log insulation score',\n",
    "                'Zygonema log insulation score',\n",
    "                'Pachynema log insulation score',\n",
    "                'H3K4me3 chromatin',\n",
    "                'H3K4me1/3 chromatin',\n",
    "                'H3K4me1 chromatin',\n",
    "                'H3K4me1+K36me3 chromatin',\n",
    "                'H3K36me3 chromatin',\n",
    "                'unmarked chromatin',\n",
    "                'H3K27me3 chromatin',\n",
    "                'Meiotic RAD21L ChIP-seq',\n",
    "                'Meiotic CTCF ChIP-seq',\n",
    "                'Meiotic RNAPII ChIP-seq',\n",
    "                'ES RAD21 ChIP-seq',\n",
    "                'ES CTCF ChIP-seq',\n",
    "                'ES RNAPII ChIP-seq']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hotspotLocs=list(set().union(peakIndsDict['smagulova2016_B6xCAST_DSB_fraglen1000_peaks'],peakIndsDict['Baker2015_prdm9_B6xCAST_peaks']))\n",
    "hotspotDf=signalDf.iloc[hotspotLocs,:]\n",
    "DSBdf=signalDf.iloc[peakIndsDict['smagulova2016_B6xCAST_DSB_fraglen1000_peaks'],:]\n",
    "prdm9Df=signalDf.iloc[peakIndsDict['Baker2015_prdm9_B6xCAST_peaks'],:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "avgPredictorList=[predictor+'_roll101mean' for predictor in predictorList]\n",
    "avgSimpleVarNames=[simpleVarName+' 500kb avg' for simpleVarName in simpleVarNames]\n",
    "fullPredictorList=predictorList+avgPredictorList\n",
    "fullSimpleVarNames=simpleVarNames+avgSimpleVarNames\n",
    "simpleDict=dict(zip(fullPredictorList,fullSimpleVarNames))\n",
    "correlationDf=hotspotDf.loc[:,fullPredictorList]\n",
    "correlationDf.rename(simpleDict, axis=1, inplace=True)\n",
    "plotCorrMatrix(correlationDf, corrMethod='spearman', saveName=f'{saveDir}/correlations.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictorList2=['Baker2015_prdm9_B6xCAST',\n",
    "               'smagulova2016_B6xCAST_DSB_fraglen1000',\n",
    "               'yin2019_crossoversCast1C_binned_weightedScore_medianNormalized',\n",
    "            'chromHMMstate_5_binned',]\n",
    "simpleVarNames2=['PRDM9 ChIP-seq',\n",
    "                'DMC1-SSDS ChIP-seq',\n",
    "                'Crossover score',\n",
    "                'H3K36me3 chromatin',]\n",
    "simpleDict2=dict(zip(predictorList2,simpleVarNames2))\n",
    "\n",
    "correlationDf2=hotspotDf.loc[:,predictorList2]\n",
    "correlationDf2.rename(simpleDict2, axis=1, inplace=True)\n",
    "plotCorrMatrix(correlationDf2, corrMethod='spearman', saveName=f'{saveDir}/recombH3K36me3corrsHotspotsOnly.pdf')\n",
    "\n",
    "correlationDf3=signalDf.loc[:,predictorList2]\n",
    "correlationDf3.rename(simpleDict2, axis=1, inplace=True)\n",
    "plotCorrMatrix(correlationDf3, corrMethod='spearman', saveName=f'{saveDir}/recombH3K36me3corrsGenomewide.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictors=' + '.join(fullPredictorList)\n",
    "output='yin2019_crossoversCast1C_binned_weightedScore_medianNormalized'\n",
    "formula=f'{output} ~ {predictors}'\n",
    "\n",
    "y, X = dmatrices(formula, data=hotspotDf, return_type='matrix')\n",
    "\n",
    "numVars=X.shape[1]-1\n",
    "\n",
    "scaler = StandardScaler()\n",
    "pca=PCA(n_components=numVars)\n",
    "varIndex=X.design_info.column_names[1:]\n",
    "PCindex=[f'PC{ind}' for ind in range(1,numVars+1)]\n",
    "X2=np.concatenate((X[:,0,None], pca.fit_transform(scaler.fit_transform(X[:,1:]))), axis=1)\n",
    "plt.rcParams.update({'font.size': 16})\n",
    "\n",
    "fig=plt.figure(figsize=(6,0.25*numVars))\n",
    "ax1=sns.barplot(y=PCindex, x=pca.explained_variance_, palette=[[0.5,0.5,0.5]])\n",
    "ax1.set_title('PCA percent explained variance')\n",
    "plt.savefig(f'{saveDir}/PCAexplainedVariance.pdf')\n",
    "print(pca.explained_variance_)\n",
    "plt.rcParams.update({'font.size': 24})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "components=pd.DataFrame(pca.transform(np.identity(numVars)), index=varIndex, columns=PCindex)\n",
    "components.rename(simpleDict, axis=0, inplace=True)\n",
    "\n",
    "componentsDf=components.unstack().reset_index()\n",
    "componentsDf['variable']=[varName.rsplit(' 500kb',1)[0] for varName in componentsDf.level_1]\n",
    "componentsDf['window']=['500kb avg' if '500kb' in varName else 'local 5kb bin' for varName in componentsDf.level_1]\n",
    "componentsDf.rename({0:'score','level_0':'PCind'}, axis=1, inplace=True)\n",
    "g=sns.catplot(y='score', x='variable', hue='window', col='PCind',col_wrap=5,kind='bar', data=componentsDf, \n",
    "               height=2, aspect=3, sharex=True, sharey=True)\n",
    "for ax in g.axes.ravel():\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "plt.tight_layout()\n",
    "plt.savefig(f'{saveDir}/PCAallLoadings.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "components=pd.DataFrame(pca.transform(np.identity(numVars)), index=varIndex, columns=PCindex)\n",
    "components['varType']=[varName.rsplit('_roll',1)[0] for varName in components.index.tolist()]\n",
    "rollVarBool=[len(varName.rsplit('_roll',1)) for varName in components.index.tolist()]\n",
    "components['rollVar']=['500kb average' if rv>1.5 else 'local 5kb bin' for rv in rollVarBool]\n",
    "\n",
    "palettes=[[(1,0.5,0),(1,0.75,0.5)],\n",
    "        [(1,0,0),(1,0.5,0.5)],\n",
    "         [(0,0,1),(0.7,0.7,1)],\n",
    "         [(0.4,0.2,0.4),(0.7,0.4,0.7)]]\n",
    "\n",
    "palettes=[[(1,0.5,0),(1,0.75,0.5)],\n",
    "        [(1,0,0),(1,0.5,0.5)],\n",
    "         [(0,0,1),(0.7,0.7,1)],\n",
    "         [(0.4,0.2,0.4),(0.7,0.4,0.7)],\n",
    "         [(0.3,0.3,0.3),(0.6,0.6,0.6)]]\n",
    "\n",
    "for ind,PCname in enumerate(PCindex[:5]):\n",
    "    plt.figure(figsize=(15,2))\n",
    "    ax=sns.barplot(y=components[PCname], x=components.varType, hue=components.rollVar, palette=palettes[ind])\n",
    "    ax.set_xticklabels(simpleVarNames)\n",
    "    ax.set_xlabel(\"\")\n",
    "    ax.set_ylabel(PCname+\" loadings\")\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)\n",
    "    plt.savefig(f'{saveDir}/{PCname}.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictors=' + '.join(['start','smagulova2016_B6xCAST_DSB_fraglen1000'])\n",
    "output='yin2019_crossoversCast1C_binned_weightedScore_medianNormalized'\n",
    "formula=f'{output} ~ {predictors}'\n",
    "\n",
    "y, X = dmatrices(formula, data=DSBdf, return_type='matrix')\n",
    "model=sm.OLS(y,X)\n",
    "results=model.fit()\n",
    "print(results.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictors=' + '.join(['start','smagulova2016_B6xCAST_DSB_fraglen1000']+fullPredictorList)\n",
    "output='yin2019_crossoversCast1C_binned_weightedScore_medianNormalized'\n",
    "formula=f'{output} ~ {predictors}'\n",
    "\n",
    "y, X = dmatrices(formula, data=DSBdf, return_type='matrix')\n",
    "\n",
    "X2=np.concatenate((X[:,0,None], scaler.fit_transform(X[:,1:3]), pca.transform(scaler.fit_transform(X[:,3:]))), axis=1)\n",
    "\n",
    "model=sm.OLS(y,X2)\n",
    "results=model.fit()\n",
    "print(results.summary())\n",
    "plt.figure(figsize=(6,0.25*len(fullPredictorList)))\n",
    "barplot=sns.barplot(y=['start','smagulova2016_B6xCAST_DSB_fraglen1000']+PCindex, x=results.tvalues[1:])\n",
    "if (type(model).__name__=='Logit'):\n",
    "    threshold=2.58\n",
    "elif (type(model).__name__=='OLS'):\n",
    "    threshold=2.33\n",
    "plt.axvline(threshold, color='k', linewidth=1, linestyle='--')\n",
    "plt.axvline(-threshold, color='k', linewidth=1, linestyle='--')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sigThreshold=0.001/X2.shape[1] #bonferroni\n",
    "varNames=['Intercept','start','smagulova2016_B6xCAST_DSB_fraglen1000'] + PCindex\n",
    "\n",
    "backwardSelected=OLSBackwardSelect(X2, y, sigThreshold)\n",
    "backwardSelectedPCnames=[varName for idx,varName in enumerate(varNames) if backwardSelected[idx]]\n",
    "\n",
    "forwardSelected=OLSForwardSelect(X2, y, sigThreshold)\n",
    "forwardSelectedPCnames=[varName for idx,varName in enumerate(varNames) if forwardSelected[idx]]\n",
    "print(forwardSelectedPCnames,backwardSelectedPCnames)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model=sm.OLS(y,X2[:, forwardSelected])\n",
    "results=model.fit()\n",
    "\n",
    "plt.figure(figsize=(6,0.35*sum(forwardSelected)))\n",
    "palette=2*[(0.6,0.6,0.6)]+[(1,0.5,0),\n",
    "         (1,0,0),\n",
    "         (0,0,1),\n",
    "         (0.4,0.2,0.4)\n",
    "        ]+len(forwardSelected)*[(0.6,0.6,0.6)]\n",
    "ax=sns.barplot(y=forwardSelectedPCnames[1:], x=results.tvalues[1:], palette=palette)\n",
    "simpleAxisLabels=['Chromosomal position', 'DMC1 ChIP-seq score'] + [ticklabel.get_text() for ticklabel in ax.get_yticklabels()[2:]]\n",
    "ax.set_yticklabels(simpleAxisLabels)\n",
    "ax.set_xlabel('t-statistic')\n",
    "threshold=stats.t.ppf(sigThreshold/2, df=X2.shape[0]-X2.shape[1]-1)\n",
    "plt.axvline(threshold, color='k', linewidth=1, linestyle='--', label='Bonferroni-corrected \\np<0.001 threshold')\n",
    "plt.axvline(-threshold, color='k', linewidth=1, linestyle='--')\n",
    "ax.legend()\n",
    "plt.savefig(f'{saveDir}/CO_results.pdf')\n",
    "print(results.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictors=' + '.join(['Baker2015_prdm9_B6xCAST'])\n",
    "output='smagulova2016_B6xCAST_DSB_fraglen1000'\n",
    "formula=f'{output} ~ {predictors}'\n",
    "\n",
    "y, X = dmatrices(formula, data=prdm9Df, return_type='matrix')\n",
    "\n",
    "model=sm.OLS(y,X)\n",
    "results=model.fit()\n",
    "print(results.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictors=' + '.join(['start','Baker2015_prdm9_B6xCAST']+fullPredictorList)\n",
    "output='smagulova2016_B6xCAST_DSB_fraglen1000'\n",
    "formula=f'{output} ~ {predictors}'\n",
    "\n",
    "y, X = dmatrices(formula, data=prdm9Df, return_type='matrix')\n",
    "\n",
    "X2=np.concatenate((X[:,0,None], scaler.fit_transform(X[:,1:3]), pca.transform(scaler.fit_transform(X[:,3:]))), axis=1)\n",
    "\n",
    "model=sm.OLS(y,X2)\n",
    "results=model.fit()\n",
    "plt.figure(figsize=(6,0.25*len(fullPredictorList)))\n",
    "barplot=sns.barplot(y=['start','grey2017_prdm9CAST_binned_maxScore']+PCindex, x=results.tvalues[1:])\n",
    "if (type(model).__name__=='Logit'):\n",
    "    threshold=2.58\n",
    "elif (type(model).__name__=='OLS'):\n",
    "    threshold=2.33\n",
    "plt.axvline(threshold, color='k', linewidth=1, linestyle='--')\n",
    "plt.axvline(-threshold, color='k', linewidth=1, linestyle='--')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sigThreshold=0.001/X2.shape[1] #bonferroni\n",
    "varNames=['Intercept','start','Baker2015_prdm9_B6xCAST'] + PCindex\n",
    "\n",
    "backwardSelected=OLSBackwardSelect(X2, y, sigThreshold)\n",
    "backwardSelectedPCnames=[varName for idx,varName in enumerate(varNames) if backwardSelected[idx]]\n",
    "\n",
    "forwardSelected=OLSForwardSelect(X2, y, sigThreshold)\n",
    "forwardSelectedPCnames=[varName for idx,varName in enumerate(varNames) if forwardSelected[idx]]\n",
    "print(forwardSelectedPCnames,backwardSelectedPCnames)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model=sm.OLS(y,X2[:, forwardSelected])\n",
    "results=model.fit()\n",
    "plt.figure(figsize=(6,0.35*sum(forwardSelected)))\n",
    "palette=1*[(0.6,0.6,0.6),(0.6,0.6,0.6)]+[(1,0.5,0)]+len(forwardSelected)*[(0.6,0.6,0.6)]\n",
    "ax=sns.barplot(y=forwardSelectedPCnames[1:], x=results.tvalues[1:], palette=palette)\n",
    "simpleAxisLabels=['Chromosomal position','PRDM9 ChIP-seq score'] + [ticklabel.get_text() for ticklabel in ax.get_yticklabels()[2:]]\n",
    "ax.set_yticklabels(simpleAxisLabels)\n",
    "ax.set_xlabel('t-statistic')\n",
    "threshold=stats.t.ppf(sigThreshold/2, df=X2.shape[0]-X2.shape[1]-1)\n",
    "plt.axvline(threshold, color='k', linewidth=1, linestyle='--', label='Bonferroni-corrected \\np<0.001 threshold')\n",
    "plt.axvline(-threshold, color='k', linewidth=1, linestyle='--')\n",
    "ax.legend()\n",
    "plt.savefig(f'{saveDir}/DSB_results.pdf')\n",
    "print(results.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictors=' + '.join(['start','Baker2015_prdm9_B6xCAST']+fullPredictorList)\n",
    "output='yin2019_crossoversCast1C_binned_weightedScore_medianNormalized'\n",
    "formula=f'{output} ~ {predictors}'\n",
    "\n",
    "y, X = dmatrices(formula, data=prdm9Df, return_type='matrix')\n",
    "\n",
    "X2=np.concatenate((X[:,0,None], scaler.fit_transform(X[:,1:3]), pca.transform(scaler.fit_transform(X[:,3:]))), axis=1)\n",
    "\n",
    "model=sm.OLS(y,X2)\n",
    "results=model.fit()\n",
    "plt.figure(figsize=(6,0.25*len(fullPredictorList)))\n",
    "barplot=sns.barplot(y=['start','Baker2015_prdm9_B6xCAST']+PCindex, x=results.tvalues[1:])\n",
    "if (type(model).__name__=='Logit'):\n",
    "    threshold=2.58\n",
    "elif (type(model).__name__=='OLS'):\n",
    "    threshold=2.33\n",
    "plt.axvline(threshold, color='k', linewidth=1, linestyle='--')\n",
    "plt.axvline(-threshold, color='k', linewidth=1, linestyle='--')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sigThreshold=0.001/X2.shape[1] #bonferroni\n",
    "varNames=['Intercept','start','Baker2015_prdm9_B6xCAST'] + PCindex\n",
    "\n",
    "backwardSelected=OLSBackwardSelect(X2, y, sigThreshold)\n",
    "backwardSelectedPCnames=[varName for idx,varName in enumerate(varNames) if backwardSelected[idx]]\n",
    "\n",
    "forwardSelected=OLSForwardSelect(X2, y, sigThreshold)\n",
    "forwardSelectedPCnames=[varName for idx,varName in enumerate(varNames) if forwardSelected[idx]]\n",
    "print(forwardSelectedPCnames,backwardSelectedPCnames)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model=sm.OLS(y,X2[:, forwardSelected])\n",
    "results=model.fit()\n",
    "plt.figure(figsize=(6,0.35*sum(forwardSelected)))\n",
    "palette=2*[(0.6,0.6,0.6)]+[(1,0,0),\n",
    "         (0,0,1),\n",
    "         (0.4,0.2,0.4)\n",
    "        ]+len(forwardSelected)*[(0.6,0.6,0.6)]\n",
    "ax=sns.barplot(y=forwardSelectedPCnames[1:], x=results.tvalues[1:], palette=palette)\n",
    "simpleAxisLabels=['Chromosomal position','PRDM9 ChIP-seq score'] + [ticklabel.get_text() for ticklabel in ax.get_yticklabels()[2:]]\n",
    "ax.set_yticklabels(simpleAxisLabels)\n",
    "ax.set_xlabel('t-statistic')\n",
    "threshold=stats.t.ppf(sigThreshold/2, df=X2.shape[0]-X2.shape[1]-1)\n",
    "plt.axvline(threshold, color='k', linewidth=1, linestyle='--', label='Bonferroni-corrected \\np<0.001 threshold')\n",
    "plt.axvline(-threshold, color='k', linewidth=1, linestyle='--')\n",
    "ax.legend()\n",
    "plt.savefig(f'{saveDir}/PRDM9toCO_results.pdf')\n",
    "print(results.summary())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
