{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cooltools import snipping\n",
    "import cooler\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pybedtools\n",
    "import bioframe\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import glob\n",
    "import os\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 24})\n",
    "chromsizes=bioframe.fetch_chromsizes('mm10')\n",
    "chromosomes = list(chromsizes.index)\n",
    "supports = [(chrom, 0, chromsizes[chrom]) for chrom in chromosomes]\n",
    "\n",
    "binsDf=pd.read_csv('../data/genomeInfo/mm10bin5kb.bed',sep='\\t',header=None,names=['chrom','start','end']) #bin file can be made with cooler makebins, binsize = 5000\n",
    "\n",
    "signalFiles=glob.glob('../data/binnedBedgraphs/*')\n",
    "signalDf=binsDf.copy()\n",
    "for signalFile in signalFiles:\n",
    "    signalDf=signalDf.merge(pd.read_csv(signalFile,sep=\"\\t\", header=None,names=['chrom','start','end',os.path.basename(signalFile).split('.')[0]]),how='left',on=['chrom','start','end'])\n",
    "    signalDf.drop_duplicates(subset=['chrom','start','end'],inplace=True)\n",
    "signalDf['zygo_inter_intra_homologRatio']=signalDf.zygonema_lucasHap_homo_cisCov/(signalDf.zygonema_lucasHap_b6_cisCov+signalDf.zygonema_lucasHap_cast_cisCov)\n",
    "signalDf['pachy_inter_intra_homologRatio']=signalDf.pachynema_lucasHap_homo_cisCov/(signalDf.pachynema_lucasHap_b6_cisCov+signalDf.pachynema_lucasHap_cast_cisCov)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def splitQuantile(toSplit,splitOn,quantiles=4):\n",
    "    quantileLims=np.linspace(0,len(toSplit),quantiles+1).astype(int)\n",
    "    sortedArray=splitOn[toSplit].sort_values().index.to_numpy()\n",
    "    quantiles=[sortedArray[quantileLims[ind]:quantileLims[ind+1]] for ind in range(quantiles)]\n",
    "    return quantiles\n",
    "\n",
    "def getPeakInds(peakFile,binsize=5000,flank=300000,supports=supports):\n",
    "    peakDf=pd.read_csv(peakFile,header=None,sep=\"\\t\",names=['chrom','start','end','name','score','strand','signalValue','pvalue','qvalue','peak'])\n",
    "    windows = snipping.make_bin_aligned_windows(\n",
    "        binsize,\n",
    "        peakDf['chrom'],\n",
    "        (peakDf['start'] + peakDf['end'])//2,\n",
    "        flank_bp=flank)\n",
    "    windows=windows.merge(pd.DataFrame(supports,columns=['chrom','zero','chromsize']),how='left',on='chrom')\n",
    "    windows=windows[(windows.end<windows.chromsize) & (windows.start>0)].drop_duplicates()\n",
    "    windows=windows.loc[(windows.chrom!='chrX')&(windows.chrom!='chrY')&(windows.chrom!='chrM'),:]\n",
    "    startInds=binsDf.reset_index().merge(windows[['chrom','start']],how='inner',on=['chrom','start'])['index']\n",
    "    endInds=binsDf.reset_index().merge(windows[['chrom','end']],how='inner',on=['chrom','end'])['index']\n",
    "    indDf=pd.DataFrame({'startInd':startInds,'endInd':endInds})\n",
    "    return ((indDf.startInd+indDf.endInd)//2).values\n",
    "\n",
    "def makeTracePlots(posListDict, signalLabelsDf, rowOrder=None,colOrder=None,palette=None, numBins=121, saveName=None, df=signalDf, avgType='median', xlabel='Relative position (kb)',aspectRatio=1.5, plotHeight=2.7,shareColAxes='stretch',symmetrize=False):\n",
    "    if rowOrder is None:\n",
    "        rowOrder=signalLabelsDf.row.unique()\n",
    "    if colOrder is None:\n",
    "        colOrder=signalLabelsDf.col.unique()\n",
    "    if len(rowOrder)==1:\n",
    "        plotHeight=plotHeight*1.25\n",
    "        aspectRatio=aspectRatio*0.8\n",
    "    halfBins=int((numBins-1)/2)\n",
    "    windowAverageDf=pd.DataFrame([])\n",
    "    genomeAverages=np.zeros(len(signalLabelsDf.index))\n",
    "    for varInd in signalLabelsDf.index:\n",
    "        var=signalLabelsDf.loc[varInd,'var']\n",
    "        colLabel=signalLabelsDf.loc[varInd,'col']\n",
    "        rowLabel=signalLabelsDf.loc[varInd,'row']\n",
    "        if avgType=='mean':\n",
    "            genomeAverages[varInd]=df[var].mean()\n",
    "        elif avgType=='median':\n",
    "            genomeAverages[varInd]=df[var].median()\n",
    "        for posListInd,posListKey in enumerate(posListDict):\n",
    "            posList=posListDict[posListKey]\n",
    "            windowAverages=np.zeros((numBins,len(posList)))\n",
    "            for posInd,pos in enumerate(posList):\n",
    "                if symmetrize:\n",
    "                    windowAverages[:,posInd]=0.5*df[var][pos-halfBins:pos+halfBins+1].values+0.5*np.flip(df[var][pos-halfBins:pos+halfBins+1].values)\n",
    "                else:\n",
    "                    windowAverages[:,posInd]=df[var][pos-halfBins:pos+halfBins+1]\n",
    "            windowDf=pd.DataFrame(windowAverages.T).stack().reset_index()\n",
    "            windowDf.rename(columns={0: \"Score\"}, inplace=True)\n",
    "            windowDf[xlabel]=5*(windowDf.level_1-halfBins)\n",
    "            windowDf['ChIP-seq peaks']=posListKey\n",
    "            windowDf['col']=colLabel\n",
    "            windowDf['row']=rowLabel\n",
    "            windowAverageDf=windowAverageDf.append(windowDf)\n",
    "    signalLabelsDf['genomeAverages']=genomeAverages\n",
    "    sharey=\"col\"\n",
    "    g=sns.relplot(x=xlabel,y=\"Score\",data=windowAverageDf, col=\"col\", row='row',hue=\"ChIP-seq peaks\", \n",
    "                  row_order=rowOrder,col_order=colOrder,\n",
    "                  ci=95, kind=\"line\", palette=palette, height=plotHeight, aspect=aspectRatio, \n",
    "                  facet_kws={'sharex':False,'sharey':\"row\" if shareColAxes=='yes' else False,'margin_titles':True})\n",
    "    yMaxSubAvg=np.zeros((len(rowOrder),len(colOrder)))\n",
    "    yMinSubAvg=np.zeros((len(rowOrder),len(colOrder)))\n",
    "    for rowInd,row in enumerate(rowOrder):\n",
    "        for colInd,col in enumerate(colOrder):\n",
    "            genomeAverage=signalLabelsDf.loc[(signalLabelsDf.row==row)&(signalLabelsDf.col==col),'genomeAverages'].values[0]\n",
    "            g.facet_axis(rowInd,colInd).axhline(y=genomeAverage, linestyle='--', color='grey', label='genome-wide '+avgType)\n",
    "            yMaxSubAvg[rowInd,colInd]=g.facet_axis(rowInd,colInd).get_ylim()[1]-genomeAverage\n",
    "            yMinSubAvg[rowInd,colInd]=g.facet_axis(rowInd,colInd).get_ylim()[0]-genomeAverage\n",
    "    yMaxSubAvgLim=np.max(yMaxSubAvg, axis=1)\n",
    "    yMinSubAvgLim=np.min(yMinSubAvg, axis=1)\n",
    "    if shareColAxes=='stretch':\n",
    "        for rowInd,row in enumerate(rowOrder):\n",
    "            for colInd,col in enumerate(colOrder):\n",
    "                genomeAverage=signalLabelsDf.loc[(signalLabelsDf.row==row)&(signalLabelsDf.col==col),'genomeAverages'].values[0]\n",
    "                g.facet_axis(rowInd,colInd).set_ylim([genomeAverage+yMinSubAvgLim[rowInd],genomeAverage+yMaxSubAvgLim[rowInd]])\n",
    "\n",
    "    plt.legend()\n",
    "    if saveName is not None:\n",
    "        plt.savefig(saveName)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peakFiles=glob.glob('../data/peakBeds/*')\n",
    "peakIndsDict=dict(zip([os.path.basename(file).split('.')[0] for file in peakFiles],[getPeakInds(file) for file in peakFiles]))\n",
    "saveDir='../outputs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "greyPrdm9castSplitonDMC=splitQuantile(peakIndsDict['grey2017orig_PRDM9_RJ2_class1_peaks'],signalDf.smagulova2016_B6xCAST_DSB_fraglen1000)\n",
    "greyPrdm9castPeakListDict={'Grey PRDM9-CAST sites (Class 1 - all)':peakIndsDict['grey2017orig_PRDM9_RJ2_class1_peaks'],\n",
    "                   'Grey PRDM9-CAST sites (top-DSB)':greyPrdm9castSplitonDMC[-1],\n",
    "                   'Grey PRDM9-CAST sites (bottom-DSB)':greyPrdm9castSplitonDMC[0]}\n",
    "greyPrdm9castPeakListMetaDict={'dict':greyPrdm9castPeakListDict,'palette':[(0.75,0.5,0.75),(0.5,0.25,0.5),(1,0.75,1)],'save':'greyprdm9cast'}\n",
    "\n",
    "greyPrdm9b6SplitonDMC=splitQuantile(peakIndsDict['grey2017orig_PRDM9_B6_class1_peaks'],signalDf.smagulova2016_B6xCAST_DSB_fraglen1000)\n",
    "greyPrdm9b6PeakListDict={'Grey PRDM9-B6 sites (Class 1 - all)':peakIndsDict['grey2017orig_PRDM9_B6_class1_peaks'],\n",
    "                   'Grey PRDM9-B6 sites (top-DSB)':greyPrdm9b6SplitonDMC[-1],\n",
    "                   'Grey PRDM9-B6 sites (bottom-DSB)':greyPrdm9b6SplitonDMC[0]}\n",
    "greyPrdm9b6PeakListMetaDict={'dict':greyPrdm9b6PeakListDict,'palette':[(0.75,0.5,0.75),(0.5,0.25,0.5),(1,0.75,1)],'save':'greyprdm9b6'}\n",
    "\n",
    "greyPrdm9allPeakListDict={'Grey PRDM9-B6 sites (Class 1)':peakIndsDict['grey2017orig_PRDM9_B6_class1_peaks'],\n",
    "                          'Grey PRDM9-B6 sites (Class 2)':peakIndsDict['grey2017orig_PRDM9_B6_class2_peaks'],\n",
    "                          'Grey PRDM9-CAST sites (Class 1)':peakIndsDict['grey2017orig_PRDM9_RJ2_class1_peaks'],\n",
    "                          'Grey PRDM9-CAST sites (Class 2)':peakIndsDict['grey2017orig_PRDM9_RJ2_class2_peaks']}\n",
    "greyPrdm9allPeakListMetaDict={'dict':greyPrdm9allPeakListDict,'palette':None,'save':'greyprdm9all'}\n",
    "\n",
    "prdm9SplitonDMC=splitQuantile(peakIndsDict['Baker2015_prdm9_B6xCAST_peaks'],signalDf.smagulova2016_B6xCAST_DSB_fraglen1000)\n",
    "prdm9PeakListDict={'PRDM9 sites (all)':peakIndsDict['Baker2015_prdm9_B6xCAST_peaks'],\n",
    "                   'PRDM9 sites (top-DSB)':prdm9SplitonDMC[-1],\n",
    "                   'PRDM9 sites (bottom-DSB)':prdm9SplitonDMC[0]}\n",
    "prdm9PeakListMetaDict={'dict':prdm9PeakListDict,'palette':[(0.75,0.5,0.75),(0.5,0.25,0.5),(1,0.75,1)],'save':'prdm9'}\n",
    "\n",
    "prdm9SplitonCO=splitQuantile(peakIndsDict['Baker2015_prdm9_B6xCAST_peaks'],signalDf.yin2019_crossoversCast1C_binned_weightedScore_medianNormalized)\n",
    "prdm9splitCOPeakListDict={'PRDM9 sites (all)':peakIndsDict['Baker2015_prdm9_B6xCAST_peaks'],\n",
    "                   'PRDM9 sites (top-CO)':prdm9SplitonCO[-1],\n",
    "                   'PRDM9 sites (bottom-CO)':prdm9SplitonCO[0]}\n",
    "prdm9splitCOPeakListMetaDict={'dict':prdm9splitCOPeakListDict,'palette':[(0.75,0.5,0.75),(0.5,0.25,0.5),(1,0.75,1)],'save':'prdm9splitCO'}\n",
    "\n",
    "DMCSplitonCO=splitQuantile(peakIndsDict['smagulova2016_B6xCAST_DSB_fraglen1000_peaks'],signalDf.yin2019_crossoversCast1C_binned_weightedScore_medianNormalized)\n",
    "DSBPeakListDict={'DSB sites (all)':peakIndsDict['smagulova2016_B6xCAST_DSB_fraglen1000_peaks'],\n",
    "                   'DSB sites (top-CO)':DMCSplitonCO[-1],\n",
    "                   'DSB sites (bottom-CO)':DMCSplitonCO[0]}\n",
    "DSBPeakListMetaDict={'dict':DSBPeakListDict,'palette':[(0.5,0.75,0.75),(0.25,0.5,0.5),(0.75,1,1)],'save':'dsb'}\n",
    "\n",
    "cohesinPeakListDict={'Meiotic RAD21L sites':peakIndsDict['vara2019_RSrad21l_peaks'],\n",
    "                          'Meiotic REC8 sites':peakIndsDict['vara2019_RSrec8_peaks'],\n",
    "                          'ES RAD21 sites':peakIndsDict['Nitzsche2011_ESC_RAD21_peaks']}\n",
    "cohesinPeakListMetaDict={'dict':cohesinPeakListDict,'palette':[(0,0.2,0.5),(0.2,0,0.5),(0.2,0.5,0.75)],'save':'cohesin'}\n",
    "\n",
    "testPeakListDict={'Meiotic RAD21L sites':peakIndsDict['vara2019_RSrad21l_peaks'][:5],\n",
    "                          'Meiotic REC8 sites':peakIndsDict['vara2019_RSrec8_peaks'][:5]}\n",
    "testPeakListMetaDict={'dict':testPeakListDict,'palette':[(0.2,0,0.5),(0.2,0.5,0.75)],'save':'test'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromHMMLabelsDf=pd.DataFrame({'var':['chromHMMstate_1_binned',\n",
    " 'chromHMMstate_2_binned',\n",
    " 'chromHMMstate_3_binned',\n",
    " 'chromHMMstate_4_binned',\n",
    " 'chromHMMstate_5_binned',\n",
    " 'chromHMMstate_6_binned',\n",
    " 'chromHMMstate_7_binned']})\n",
    "chromHMMLabelsDf['col']=[var.split('_')[1] for var in chromHMMLabelsDf['var']]\n",
    "chromHMMLabelsDf['row']=0\n",
    "chromHMMLabelsDf.index.name='chromhmm'\n",
    "\n",
    "chipLabelsDf=pd.DataFrame(['Nitzsche2011_ESC_RAD21',\n",
    "    'vara2019_PDctcf',\n",
    "    'margolin2014RNAPII16dpp',\n",
    "    'ES_RNAP2_binned_maxScore',\n",
    "    'Nitzsche2011_ESC_CTCF',\n",
    "    'vara2019_PDrad21l'],columns=['var'])\n",
    "chipLabelsDf['col']=['ES','Meiosis','Meiosis','ES','ES','Meiosis']\n",
    "chipLabelsDf['row']=['RAD21(L)','CTCF','RNAPII','RNAPII','CTCF','RAD21(L)']\n",
    "chipLabelsDf.index.name='chip'\n",
    "\n",
    "\n",
    "recombLabelsDf=pd.DataFrame(['Baker2015_prdm9_B6xCAST','smagulova2016_B6xCAST_DSB_fraglen1000','yin2019_crossoversCast1C_binned_weightedScore_medianNormalized'],columns=['var'])\n",
    "recombLabelsDf['col']=['Baker 2015 PRDM9','Smagulova 2016 DSBs','Yin 2019 Crossovers']\n",
    "recombLabelsDf['row']=0\n",
    "recombLabelsDf.index.name='recomb'\n",
    "\n",
    "tradHicSignalLabelsDf=pd.DataFrame({'var':[\n",
    "    'ES_cisTotalRatio',\n",
    "    'zygo_E1chunks',\n",
    "    'pachy_logInsScore',\n",
    "    'pachy_cisTotalRatio',\n",
    "    'zygo_fireScore',\n",
    "    'zygo_logInsScore',\n",
    "    'ES_E1chunks',\n",
    "    'pachy_E1chunks',\n",
    "    'ES_logInsScore',\n",
    "    'pachy_fireScore',\n",
    "    'ES_fireScore',\n",
    "    'zygo_cisTotalRatio']})\n",
    "tradHicSignalLabelsDf[['col','row']]=tradHicSignalLabelsDf['var'].str.split('_',1,expand=True)\n",
    "tradHicSignalLabelsDf['row']=tradHicSignalLabelsDf.row.map({'cisTotalRatio':'CTR','E1chunks':'CS','logInsScore':'LIS','fireScore':'FS'})\n",
    "tradHicSignalLabelsDf.index.name='hic'\n",
    "\n",
    "hicWangSignalLabelsDf=pd.DataFrame({'var':[\n",
    "    'ES_cisTotalRatio',\n",
    "    'zygo_E1chunks',\n",
    "    'pachy_logInsScore',\n",
    "    'pachy_cisTotalRatio',\n",
    "    'zygo_fireScore',\n",
    "    'zygo_logInsScore',\n",
    "    'ES_E1chunks',\n",
    "    'pachy_E1chunks',\n",
    "    'ES_logInsScore',\n",
    "    'pachy_fireScore',\n",
    "    'ES_fireScore',\n",
    "    'zygo_cisTotalRatio',\n",
    "    'PAC_logInsScore',\n",
    "    'PAC_fireScore',\n",
    "    'PAC_E1chunks',\n",
    "    'PAC_cisTotalRatio']})\n",
    "hicWangSignalLabelsDf[['col','row']]=hicWangSignalLabelsDf['var'].str.split('_',1,expand=True)\n",
    "hicWangSignalLabelsDf['row']=hicWangSignalLabelsDf.row.map({'cisTotalRatio':'CTR','E1chunks':'CS','logInsScore':'LIS','fireScore':'FS'})\n",
    "hicWangSignalLabelsDf.index.name='hicWang'\n",
    "\n",
    "lucasHapSignalLabelsDf=pd.DataFrame({'var':['zygonema_lucasHap_b6_cisTotalRatio',\n",
    " 'zygonema_lucasHap_cast_cisTotalRatio',\n",
    " 'zygonema_lucasHap_homo_E1chunks',\n",
    " 'pachynema_lucasHap_b6_E1chunks',\n",
    " 'pachynema_lucasHap_b6_cisTotalRatio',\n",
    " 'pachynema_lucasHap_cast_E1chunks',\n",
    " 'zygonema_lucasHap_b6_E1chunks',\n",
    " 'pachynema_lucasHap_homo_cisTotalRatio',\n",
    " 'zygonema_lucasHap_b6_cisTotalRatio',\n",
    " 'pachynema_lucasHap_cast_cisTotalRatio',\n",
    " 'pachynema_lucasHap_homo_E1chunks',\n",
    " 'zygonema_lucasHap_cast_E1chunks',\n",
    " 'zygonema_lucasHap_homo_cisTotalRatio']})\n",
    "lucasHapSignalLabelsDf[['col','row']]=lucasHapSignalLabelsDf['var'].str.rsplit('_',1,expand=True)\n",
    "lucasHapSignalLabelsDf['row']=lucasHapSignalLabelsDf.row.map({'cisTotalRatio':'CTR','E1chunks':'CS','logInsScore':'LIS','fireScore':'FS'})\n",
    "lucasHapSignalLabelsDf['col']=lucasHapSignalLabelsDf.col.map({'pachynema_lucasHap_b6':'pa_b6','pachynema_lucasHap_cast':'pa_ca','pachynema_lucasHap_homo':'pa_ho','zygonema_lucasHap_b6':'zy_b6','zygonema_lucasHap_cast':'zy_ca','zygonema_lucasHap_homo':'zy_ho'})\n",
    "lucasHapSignalLabelsDf.index.name='lucashap'\n",
    "\n",
    "extraCTRSignalLabelsDf=pd.DataFrame(['varaLZ_cisTotalRatio',\n",
    "                                    'varaPD_cisTotalRatio',\n",
    "                                    'varaRS_cisTotalRatio',\n",
    "                                    'alavattamRS_cisTotalRatio'],columns=['var'])\n",
    "extraCTRSignalLabelsDf['col']=[var.split('_')[0] for var in extraCTRSignalLabelsDf['var']]\n",
    "extraCTRSignalLabelsDf['row']=0\n",
    "extraCTRSignalLabelsDf.index.name='extraCTR'\n",
    "\n",
    "interhomologLabelsDf=pd.DataFrame({'var':['zygo_inter_intra_homologRatio','pachy_inter_intra_homologRatio']})\n",
    "interhomologLabelsDf['col']=['Inter-vs.-intra\\nhomolog ratio\\n(zygonema)','Inter-vs.-intra\\nhomolog ratio\\n(pachynema)']\n",
    "interhomologLabelsDf['row']=0\n",
    "interhomologLabelsDf.index.name='interhomolog'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peakListMetaDict=cohesinPeakListMetaDict\n",
    "for signalLabelDf in [chromHMMLabelsDf,extraCTRSignalLabelsDf,hicWangSignalLabelsDf]:\n",
    "    if signalLabelDf.index.name=='extraCTR':\n",
    "        shareColAxes='no'\n",
    "    elif signalLabelDf.index.name=='chromhmm':\n",
    "        shareColAxes='no'\n",
    "        avgType='mean'\n",
    "    else:\n",
    "        shareColAxes='stretch'\n",
    "    makeTracePlots(peakListMetaDict['dict'],signalLabelDf,\n",
    "                       palette=peakListMetaDict['palette'],\n",
    "                      saveName=f\"{saveDir}/{peakListMetaDict['save']}_{signalLabelDf.index.name}_traceplots.pdf\",\n",
    "                       xlabel='Pos. (kb)',aspectRatio=1.5,shareColAxes=shareColAxes,symmetrize=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for signalLabelDf in [lucasHapSignalLabelsDf,chromHMMLabelsDf,chipLabelsDf,tradHicSignalLabelsDf]:\n",
    "    for peakListMetaDict in [prdm9splitCOPeakListMetaDict,prdm9PeakListMetaDict,DSBPeakListMetaDict,greyPrdm9castPeakListMetaDict,greyPrdm9b6PeakListMetaDict,greyPrdm9allPeakListMetaDict]:\n",
    "        if signalLabelDf.index.name=='chromhmm':\n",
    "            shareColAxes='no'\n",
    "            avgType='mean'\n",
    "        else:\n",
    "            shareColAxes='stretch'\n",
    "            avgType='median'            \n",
    "        makeTracePlots(peakListMetaDict['dict'],signalLabelDf,\n",
    "                       palette=peakListMetaDict['palette'],\n",
    "                      saveName=f\"{saveDir}/{peakListMetaDict['save']}_{signalLabelDf.index.name}_traceplots.pdf\",\n",
    "                       xlabel='Pos. (kb)',aspectRatio=1.5,\n",
    "                      shareColAxes=shareColAxes,avgType=avgType,symmetrize=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for signalLabelDf in [interhomologLabelsDf]:\n",
    "    for peakListMetaDict in [prdm9PeakListMetaDict,DSBPeakListMetaDict]:\n",
    "        if signalLabelDf.index.name=='chromhmm':\n",
    "            shareColAxes='no'\n",
    "            avgType='mean'\n",
    "        else:\n",
    "            shareColAxes='stretch'\n",
    "            avgType='median'            \n",
    "        makeTracePlots(peakListMetaDict['dict'],signalLabelDf,\n",
    "                       palette=peakListMetaDict['palette'],\n",
    "                      saveName=f\"{saveDir}/{peakListMetaDict['save']}_{signalLabelDf.index.name}_traceplots.pdf\",\n",
    "                       xlabel='Pos. (kb)',aspectRatio=1.5,\n",
    "                      shareColAxes=shareColAxes,avgType=avgType,symmetrize=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
