{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cooltools import snipping\n",
    "import cooler\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pybedtools\n",
    "import bioframe\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import glob\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 16})\n",
    "chromsizes=bioframe.fetch_chromsizes('mm10')\n",
    "chromosomes = list(chromsizes.index)\n",
    "supports = [(chrom, 0, chromsizes[chrom]) for chrom in chromosomes]\n",
    "\n",
    "binsDf=pd.read_csv('../data/genomeInfo/mm10bin5kb.bed',sep='\\t',header=None,names=['chrom','start','end']) #bin file can be made with cooler makebins, binsize = 5000\n",
    "\n",
    "signalFiles=glob.glob('../data/binnedBedgraphs/*')\n",
    "signalDf=binsDf.copy()\n",
    "for signalFile in signalFiles:\n",
    "    signalDf=signalDf.merge(pd.read_csv(signalFile,sep=\"\\t\", header=None,names=['chrom','start','end',os.path.basename(signalFile).split('.')[0]]),how='left',on=['chrom','start','end'])\n",
    "    signalDf.drop_duplicates(subset=['chrom','start','end'],inplace=True)\n",
    "chromHMMdict={1:'H3K4me3', \n",
    "       2:'H3K4me1/3',\n",
    "       3:'H3K4me1',\n",
    "       4:'H3K4me1+\\nH3K36me3',\n",
    "       5:'H3K36me3',\n",
    "       6:'Unmarked',\n",
    "       7:'H3K27me3'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getPeakDf(peakFile):\n",
    "    return pd.read_csv(peakFile,header=None,sep=\"\\t\",names=['chrom','start','end','name','score','strand','signalValue','pvalue','qvalue','peak'])\n",
    "\n",
    "def chromhmmLabelPeakDf(peakDf,chromHmmFile='../data/chromhmm/ENCFF215ZKG_mm9mm10liftOver.bed'):\n",
    "    peakCenterDf=pd.DataFrame(peakDf['chrom'])\n",
    "    peakCenterDf['start']=(peakDf.start+peakDf.end)//2\n",
    "    peakCenterDf['end']=peakCenterDf.start+1\n",
    "    peakBt=pybedtools.bedtool.BedTool.from_dataframe(peakCenterDf)\n",
    "    chromhmmBt=pybedtools.BedTool(chromHmmFile)\n",
    "    return peakBt.intersect(chromhmmBt,wb=True).to_dataframe()[['chrom','start','end','thickStart']].rename(columns={'thickStart':'chromHMM'})\n",
    "\n",
    "def getSplitQuantileDict(posListDict,splitOn,baseLabel=\"sites\",splitLabel='CO'):\n",
    "    quantiles=splitQuantile(posListDict,splitOn)\n",
    "    return {f'{baseLabel} (all)':posListDict,\n",
    "           f'{baseLabel} (top-{splitLabel})':quantiles[-1],\n",
    "           f'{baseLabel} (bottom-{splitLabel})':quantiles[0]}\n",
    "\n",
    "def splitQuantile(toSplit,splitOn,quantiles=4):\n",
    "    quantileLims=np.linspace(0,len(toSplit),quantiles+1).astype(int)\n",
    "    sortedArray=splitOn[toSplit].sort_values().index.to_numpy()\n",
    "    quantiles=[sortedArray[quantileLims[ind]:quantileLims[ind+1]] for ind in range(quantiles)]\n",
    "    return quantiles\n",
    "\n",
    "def getPeakInds(peakFile,binsize=5000,flank=300000,supports=supports):\n",
    "    if type(peakFile) is str:\n",
    "        peakDf=pd.read_csv(peakFile,header=None,sep=\"\\t\",names=['chrom','start','end','name','score','strand','signalValue','pvalue','qvalue','peak'])\n",
    "    elif type(peakFile) is pd.core.frame.DataFrame:\n",
    "        peakDf=peakFile\n",
    "    windows = snipping.make_bin_aligned_windows(\n",
    "        binsize,\n",
    "        peakDf['chrom'],\n",
    "        (peakDf['start'] + peakDf['end'])//2,\n",
    "        flank_bp=flank)\n",
    "    windows=windows.merge(pd.DataFrame(supports,columns=['chrom','zero','chromsize']),how='left',on='chrom')\n",
    "    windows=windows[(windows.end<windows.chromsize) & (windows.start>0)].drop_duplicates()\n",
    "    windows=windows.loc[(windows.chrom!='chrX')&(windows.chrom!='chrY')&(windows.chrom!='chrM'),:]\n",
    "    startInds=binsDf.reset_index().merge(windows[['chrom','start']],how='inner',on=['chrom','start'])['index']\n",
    "    endInds=binsDf.reset_index().merge(windows[['chrom','end']],how='inner',on=['chrom','end'])['index']\n",
    "    indDf=pd.DataFrame({'startInd':startInds,'endInd':endInds})\n",
    "    return ((indDf.startInd+indDf.endInd)//2).values\n",
    "\n",
    "def makeTracePlots(posListDict, signalLabelsDf, rowOrder=None,colOrder=None,palette=None, numBins=121, saveName=None, df=signalDf, avgType='median', xlabel='Relative position (kb)',aspectRatio=1.2, shareColAxes='stretch',titleString=None):\n",
    "    if rowOrder is None:\n",
    "        rowOrder=signalLabelsDf.row.unique()\n",
    "    if colOrder is None:\n",
    "        colOrder=signalLabelsDf.col.unique()\n",
    "    halfBins=int((numBins-1)/2)\n",
    "    windowAverageDf=pd.DataFrame([])\n",
    "    genomeAverages=np.zeros(len(signalLabelsDf.index))\n",
    "    for varInd in signalLabelsDf.index:\n",
    "        var=signalLabelsDf.loc[varInd,'var']\n",
    "        colLabel=signalLabelsDf.loc[varInd,'col']\n",
    "        rowLabel=signalLabelsDf.loc[varInd,'row']\n",
    "        if avgType=='mean':\n",
    "            genomeAverages[varInd]=df[var].mean()\n",
    "        elif avgType=='median':\n",
    "            genomeAverages[varInd]=df[var].median()\n",
    "        for posListInd,posListKey in enumerate(posListDict):\n",
    "            posList=posListDict[posListKey]\n",
    "            windowAverages=np.zeros((numBins,len(posList)))\n",
    "            for posInd,pos in enumerate(posList):\n",
    "                windowAverages[:,posInd]=df[var][pos-halfBins:pos+halfBins+1]\n",
    "            windowDf=pd.DataFrame(windowAverages.T).stack().reset_index()\n",
    "            windowDf.rename(columns={0: \"Score\"}, inplace=True)\n",
    "            windowDf[xlabel]=5*(windowDf.level_1-halfBins)\n",
    "            windowDf['ChIP-seq peaks']=posListKey\n",
    "            windowDf['col']=colLabel\n",
    "            windowDf['row']=rowLabel\n",
    "            windowAverageDf=windowAverageDf.append(windowDf)\n",
    "    signalLabelsDf['genomeAverages']=genomeAverages\n",
    "    g=sns.relplot(x=xlabel,y=\"Score\",data=windowAverageDf, col=\"col\", row='row',hue=\"ChIP-seq peaks\", \n",
    "                  row_order=rowOrder,col_order=colOrder,\n",
    "                  ci=95, kind=\"line\", palette=palette, height=2.7, aspect=aspectRatio, \n",
    "                  facet_kws={'sharey':False,'sharex':(shareColAxes=='yes'),'margin_titles':True})\n",
    "    yMaxSubAvg=np.zeros((len(rowOrder),len(colOrder)))\n",
    "    yMinSubAvg=np.zeros((len(rowOrder),len(colOrder)))\n",
    "    for rowInd,row in enumerate(rowOrder):\n",
    "        for colInd,col in enumerate(colOrder):\n",
    "            genomeAverage=signalLabelsDf.loc[(signalLabelsDf.row==row)&(signalLabelsDf.col==col),'genomeAverages'].values[0]\n",
    "            g.facet_axis(rowInd,colInd).axhline(y=genomeAverage, linestyle='--', color='grey', label='genome-wide '+avgType)\n",
    "            yMaxSubAvg[rowInd,colInd]=g.facet_axis(rowInd,colInd).get_ylim()[1]-genomeAverage\n",
    "            yMinSubAvg[rowInd,colInd]=g.facet_axis(rowInd,colInd).get_ylim()[0]-genomeAverage\n",
    "    yMaxSubAvgLim=np.max(yMaxSubAvg, axis=1)\n",
    "    yMinSubAvgLim=np.min(yMinSubAvg, axis=1)\n",
    "    if shareColAxes=='stretch':\n",
    "        for rowInd,row in enumerate(rowOrder):\n",
    "            for colInd,col in enumerate(colOrder):\n",
    "                genomeAverage=signalLabelsDf.loc[(signalLabelsDf.row==row)&(signalLabelsDf.col==col),'genomeAverages'].values[0]\n",
    "                g.facet_axis(rowInd,colInd).set_ylim([genomeAverage+yMinSubAvgLim[rowInd],genomeAverage+yMaxSubAvgLim[rowInd]])\n",
    "\n",
    "#     g.fig.tight_layout()\n",
    "\n",
    "    plt.legend()\n",
    "    if titleString is not None:\n",
    "        plt.title(titleString)\n",
    "    if saveName is not None:\n",
    "        plt.savefig(saveName)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peakFiles=glob.glob('../data/peakBeds/*')\n",
    "peakIndsDict=dict(zip([os.path.basename(file).split('.')[0] for file in peakFiles],[getPeakInds(file) for file in peakFiles]))\n",
    "peakDfDict=dict(zip([os.path.basename(file).split('.')[0] for file in peakFiles],[getPeakDf(file) for file in peakFiles]))\n",
    "saveDir='../outputs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pachyCTRsignalLabelsDf=pd.DataFrame({'var':['pachy_cisTotalRatio'],'row':[0],'col':[0]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohesinChromhmmLabelsDict={'Meiotic RAD21L':chromhmmLabelPeakDf(peakDfDict['vara2019_PDrad21l_peaks']),\n",
    "                  'Meiotic REC8':chromhmmLabelPeakDf(peakDfDict['vara2019_PDrec8_peaks']),\n",
    "                  'ES RAD21':chromhmmLabelPeakDf(peakDfDict['Nitzsche2011_ESC_RAD21_peaks'])}\n",
    "cohesinChromhmmDfDict=dict([(key+\":\\n\"+chromHMMdict[chromHMMstate],cohesinChromhmmLabelsDict[key][cohesinChromhmmLabelsDict[key].chromHMM==chromHMMstate][['chrom','start','end']]) for key in cohesinChromhmmLabelsDict for chromHMMstate in cohesinChromhmmLabelsDict[key].chromHMM.unique()])\n",
    "cohesinChromhmmDfDictList=[dict([(key,cohesinChromhmmDfDict[key]) for key in cohesinChromhmmDfDict.keys() if key.split(\":\\n\")[1]==chromHMMstate]) for chromHMMstate in chromHMMdict.values()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cohesinChromhmmDfDict in cohesinChromhmmDfDictList:\n",
    "    peakIndsDict=dict([(key.split(\":\\n\")[0],getPeakInds(cohesinChromhmmDfDict[key])) for key in cohesinChromhmmDfDict])\n",
    "    titleString=list(cohesinChromhmmDfDict.keys())[0].split(\":\\n\")[1]\n",
    "    chromHMMstate=titleString.replace(\"/\",\"_\").replace(\"\\n\",\"_\")\n",
    "    makeTracePlots(peakIndsDict, pachyCTRsignalLabelsDf,aspectRatio=1.5, palette=[(0,0.2,0.5),(0.2,0,0.5),(0.2,0.5,0.75)],\n",
    "                   saveName=f\"{saveDir}/cohesinsChromhmm_{chromHMMstate}_pachyCTRtraceplot.pdf\",titleString=titleString)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DSBpeakChromHMMDf=chromhmmLabelPeakDf(peakDfDict['smagulova2016_B6xCAST_DSB_fraglen1000_peaks'])\n",
    "DSBpeakChromHMMDfDict=dict([(chromHMMdict[chromHMMstate],DSBpeakChromHMMDf[DSBpeakChromHMMDf.chromHMM==chromHMMstate][['chrom','start','end']]) for chromHMMstate in DSBpeakChromHMMDf.chromHMM.unique()])\n",
    "DSBpeakChromHMMPeakIndsDictCOsplit=dict([(key,getSplitQuantileDict(posListDict=getPeakInds(DSBpeakChromHMMDfDict[key]),splitOn=signalDf.yin2019_crossoversCast1C_binned_weightedScore_medianNormalized,baseLabel=\"DSB sites\",splitLabel='CO')) for key in DSBpeakChromHMMDfDict])\n",
    "for key in DSBpeakChromHMMPeakIndsDictCOsplit:\n",
    "    chromHMMstate=key.replace(\"/\",\"_\").replace(\"\\n\",\"_\")\n",
    "    makeTracePlots(DSBpeakChromHMMPeakIndsDictCOsplit[key], pachyCTRsignalLabelsDf,aspectRatio=1.5, palette=[(0.5,0.75,0.75),(0.25,0.5,0.5),(0.75,1,1)],\n",
    "                   saveName=f\"{saveDir}/DSBsplitCO_chromhmm_{chromHMMstate}_pachyCTRtraceplot.pdf\",titleString=key)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
